addons/web/static/src/js/framework/data.js

```diff
@@ -737,43 +739,123 @@ var BufferedDataSet = DataSetStatic.extend({
             self.last_default_get = res;
         });
     },
+    _update_cache: function (id, options) {
+        var cached = _.detect(this.cache, function(x) {return x.id === id;});
```

Comment:
one should call this method after modifying `this.from_read`, `this.to_create` or `this.to_write`, and updates `this.cache` and `this.from_readonly`